### PR TITLE
api: do not allow FQDNSelectors to contain both MatchName and MatchPattern

### PR DESCRIPTION
**Pull request by ianvernon**
Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7963)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
I'm OK with this change, as it seems like the behaviour has actually been effectively the same as having two independent fqdn rules with either of the fields set, which means:
* Users can easily migrate their ruleset if necessary;
* Also this is somewhat inconsistent with other fields embedded at the same level of part of a policy

It's unlikely that many users put both in the same rule anyway. I don't think we explicitly documented this to state the expected behaviour in this case either.

**Comment by ianvernon** [comment]
From offline discussion - add documentation for upgrade guide.

**Review by nebril — APPROVED** [review]

**Review by vadorovsky — APPROVED** [review]

File: pkg/policy/api/fqdn.go [hunk 1]
```diff
@@ -62,6 +62,9 @@ func (s *FQDNSelector) String() string {
 // when using MatchName the basic requirement is that is a valid regexp. We
 // test that it can compile here.
 func (s *FQDNSelector) sanitize() error {
+	if len(s.MatchName) > 0 && len(s.MatchPattern) > 0 {
+		return fmt.Errorf("only one of MatchName or MatchPattern is allowed in an FQDNSelector")
```
Comment (aanm) [inline pkg/policy/api/fqdn.go:66]: This is surprising to me.. Is there a reason why we can't have both? cc @raybejjani

**Review by aanm — APPROVED** [review]

File: pkg/policy/api/fqdn.go (hunk 1 above)
Comment (ianvernon) [reply pkg/policy/api/fqdn.go:66]: The rationale behind this is that with the upcoming usage of `SelectorCache`, we will have a mapping of a selector --> identities which match that selector. In the case of FQDN, the fact that the same selector can have two disjoint fields which may match completely separate things does not make sense conceptually - say I have `MatchName: google.com` and then `MatchPattern: *.cilium.io` - both are completely disjoint. I want to be able to map a given FQDNSelector to the identities which match that selector - if we have two fields which are disjoint in the same selector, we lose a lot of granularity. Also, I will be using the `FQDNSelector` as a key in a map and mapping it to a regular expression representing the selector, so having both fields populated would make such behavior impossible.

**Review by James203 — COMMENTED** [review]
Duplicate of #

**Comment by ianvernon** [comment]
@James203 what is this a duplicate of?

**Pull request merged**